pymc_marketing/clv/models/shifted_beta_geo.py

```diff
@@ -0,0 +1,203 @@
+
+            import pymc as pm
+            from pymc_marketing.clv import ShiftedBetaGeoModelIndividual
+
+            model = ShiftedBetaGeoModelIndividual(
+                customer_id=[0, 1, 2, 3, ...],
+                t_churn=[1, 2, 8, 4, 8 ...],
+                T=8,  # Can also be an array with one value per customer
+                alpha_prior=pm.HalfNormal.dist(10),
+                beta_prior=pm.HalfNormal.dist(10),
+            )
+
+            model.fit()
+            print(model.fit_summary())
+
+            # Predict how many periods in the future are existing customers
+            likely to cancel (ignoring that some may already have cancelled)
+            expected_churn_time = model.distribution_customer_churn_time(
+                customer_id=[0, 1, 2, 3, ...],
+            )
+            print(expected_churn_time.mean("customer_id"))
+
+            # Predict churn time for 10 new customers, conditioned on data
+            new_customers_churn_time = model.distribution_new_customer_churn_time(n=10)
+            print(new_customers_churn_time.mean("new_customer_id"))
+
+
+    References
+    ----------
+    .. [1] Fader, P. S., & Hardie, B. G. (2007). How to project customer retention.
+           Journal of Interactive Marketing, 21(1), 76-90.
+           https://journals.sagepub.com/doi/pdf/10.1002/dir.20074
+    """
+
+    _model_name = "Shifted-Beta-Geometric Model (Individual Customers)"
+
+    def __init__(
+        self,
+        customer_id: Union[np.ndarray, pd.Series],
+        t_churn: Union[np.ndarray, pd.Series],
+        T: Union[np.ndarray, pd.Series],
+        alpha_prior: Optional[TensorVariable] = None,
+        beta_prior: Optional[TensorVariable] = None,
+    ):
+        super().__init__()
+
+        t_churn = np.asarray(t_churn)
+        T = np.asarray(T)
+
+        if np.any((t_churn < 0) | (t_churn > T) | np.isnan(t_churn)):
+            raise ValueError(
+                "t_churn must respect 0 < t_churn <= T.\n",
+                "Customers that are still alive should have t_churn = T",
+            )
+
+        alpha_prior, beta_prior = self._process_priors(alpha_prior, beta_prior)
+
+        coords = {"customer_id": np.asarray(customer_id)}
+        with pm.Model(coords=coords) as self.model:
+            alpha = self.model.register_rv(alpha_prior, name="alpha")
+            beta = self.model.register_rv(beta_prior, name="beta")
+
+            theta = pm.Beta("theta", alpha, beta, dims=("customer_id",))
+
+            churn_raw = pm.Geometric.dist(theta)
+
+            pm.Censored(
+                "churn_censored",
+                churn_raw,
+                lower=None,
+                upper=T,
+                observed=t_churn,
+                dims=("customer_id",),
+            )
+
+    def _process_priors(self, alpha_prior, beta_prior):
+        if alpha_prior is None:
+            alpha_prior = pm.HalfFlat.dist()
+        else:
+            assert alpha_prior.eval().shape == ()
```

Comment:
Shall we avoid `assert` statements and use exception handling instead and throw more informative errors? See https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html

Reply:
This gets my vote. These `asserts` will be ignored if anyone runs this code in optimized mode via `python -o script.py`. Here are some interesting links on the topic:

https://snyk.io/blog/the-dangers-of-assert-in-python/

https://discuss.python.org/t/stop-ignoring-asserts-when-running-in-optimized-mode/13132

Reply:
Sure, I'll raise informative ValueErrors instead